batchproducer/batchproducer.go

```diff
@@ -233,15 +232,22 @@ func (b *batchProducer) Stop() error {
 }
 
 func (b *batchProducer) setRunning(running bool) {
-	b.runningMu.Lock()
-	defer b.runningMu.Unlock()
-	b.running = running
+	var newValue int32
+	if running {
+		newValue = 1
+	} else {
+		newValue = 0
+	}
+
+	oldValue := b.running
+
+	for swapped := false; !swapped; {
```

Comment:
> Do we start-stop all that often to warrant this spin loop?

Not so often, but I don’t understand… `atomic.CompareAndSwapInt32` returns `(swapped bool)` which means that the operation can fail. So isn’t it correct to check the return value and try again if it failed?

Also this isn’t so much of a “spinloop” — it will exit as soon as `swapped` is `true`. If you think it’d be prudent I could put a short `sleep` in the loop.

> I think that calling start/stop while the component is already started/stopped to be a caller error and thus should yield no effect.

That is the case. In normal operation, Start and Stop aren’t called very much. But I still want to avoid data races. That’s why I was using a mutex for `running`.